### ISSUE TITLE
Jira Links not copied/mirrored in the ADO Work Item

### ISSUE THREAD

**Issue by steverwyllie** [solidify/jira-azuredevops-migrator#990]
**Describe the problem**
I'm failing to get several types of Jira Links to be copied/mirrored on the ADO Work Items created by the Migrator.
(I've attached screenshots of the links contained in a test Jira Story and the links the Migrator generated in the ADO Story)

**To Reproduce**
- I created a test Jira Story, that made use of all Jira link types
- I then used JQL in the Jira Exporter, that picked up this Story and all Jira tickets it was linked to
- I tried to define each link-type in the Migrator link-map, but I'm not confident I'm using the link-map correctly
- The ADO User Story created by the Migrator didn't have many of the links contained in the original Jira Story

**Tool version**
 - 3.0.356

**Attachments**
- config.json
[config-scrum.json](https://github.com/solidify/jira-azuredevops-migrator/files/14363600/config-scrum.json)

- jira-export-log.txt
[jira-export-log-240221-175052.txt](https://github.com/solidify/jira-azuredevops-migrator/files/14363606/jira-export-log-240221-175052.txt)

- wi-import-log.txt
[wi-import-log-240221-175113.txt](https://github.com/solidify/jira-azuredevops-migrator/files/14363610/wi-import-log-240221-175113.txt)

**Screenshots**
![image](https://github.com/solidify/jira-azuredevops-migrator/assets/4835941/b5b937d3-d0a6-4255-a876-ce0a78a92bfd)
![image](https://github.com/solidify/jira-azuredevops-migrator/assets/4835941/c79a6925-fc1f-4cc1-9698-71957bba3b58)

**Comment by Alexander-Hjelm** [comment]
Looking through your link map inside your config, I am wondering if maybe one should use the LinkType names instead of the inward/outward descriptions? Give that a try instead and let me know if that gives better results. I am attaching a screenshot to show how to get the LinkType names in Jira Cloud:

![image](https://github.com/solidify/jira-azuredevops-migrator/assets/10683896/607a976c-92d3-4eb9-90f8-7bf583880fae)

**Comment by steverwyllie** [comment]
Hi Alexander. Thanks for the tip. We're using on-premise Jira (see pic for link names). Is this the correct JSON to use? (I'm finding it difficult to judge when to use Forward vs Reverse)
	  {
        "source": "Blocks",
        "target": "System.LinkTypes.Dependency-Reverse"
      },
      {
        "source": "Cloners",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
        "source": "Duplicate",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
        "source": "Relates",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "jira_subtask_link",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Epic-Story Link",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Mother-Child",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Issue Split",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
        "source": "Parent-Child Link",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },

![image](https://github.com/solidify/jira-azuredevops-migrator/assets/4835941/f7e307ff-d562-48da-ab92-e6f75bb8f247)

**Comment by steverwyllie** [comment]
I'm very sorry; a combination of trial - and error - means I'm now getting the link types

**Comment by ankusinha23** [comment]
hi @steverwyllie .. i m also striggling with same issue and mine also is same JIRA cloud in premise .. can u share in details how u fixed the issue and also the mapping u did

**Comment by steverwyllie** [comment]
Hi @ankusinha23 I'm not sure I'm a reliable source of how to use this tool, but hope this may help (a bit)

1. The core issue seemed to be that we have 1000's of Jira tickets to migrate, so thought it best to split it into multiple migration 'runs - and I wasn't including all the Related To items in the same Migration 'run'

2. Our quite messy Jira setup also means the tool stuggled to follow the 'breadcrumb trail' of Related To items. For example, we have several examples of people making Epics children of Bugs, and I didn't even realise that was possible! (I appreciate this isn't a Related To example, but perhaps indicates how unstructured our Jira tickets have become, over many years.)

3. So I'm now manually controlling what Related To items are in the same Migration 'run'
- using "download-options": 0, in the JSON  and
- JQL "Project = xxxxx AND issuekey in (11111,22222)   OR   (  issue in linkedIssues(11111)    OR issue in linkedIssues(22222)"
- This gives me a results list I can then export to Excel, delete items I don't want migrated, then use excel text-management to construct a long sting of JQL of "issuekey in (11111,22222)"
- I then test that JQL in Jira, check Jira accepts the syntax, then copy-and-paste that JQL into the Migrator JSON file

4. I still struggle to understand all the Reverse and Forward Related To options so I'm using the following (probably wrong-headed) statements in the JSON file.
  "link-map": {
    "link": [
	  {
        "source": "Blocks",
        "target": "System.LinkTypes.Dependency-Forward"
      },
      {
        "source": "Cloners",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
        "source": "Duplicate",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
        "source": "Relates",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "jira_subtask_inward",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "Epic-Story Link",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Mother-Child",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "Issue Split",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "Parent-Child Link",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "Epic",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "Parent",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "is Epic of",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "has Epic",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "is parent of",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "is mother of",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "is child of",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "jira_subtask_outward",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "jira_subtask_inward",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
      {
        "source": "split from",
        "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
        "source": "split to",
        "target": "System.LinkTypes.Hierarchy-Forward"
      },
	  {
        "source": "blocks",
        "target": "System.LinkTypes.Dependency-Forward"
      },
	  {
        "source": "is blocked by",
        "target": "System.LinkTypes.Dependency-Reverse"
      },
      {
        "source": "duplicates",
        "target": "System.LinkTypes.Duplicate-Reverse"
      },
	  {
        "source": "is duplicated by",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
	  {
        "source": "clones",
        "target": "System.LinkTypes.Duplicate-Reverse"
      },
	  {
        "source": "is cloned by",
        "target": "System.LinkTypes.Duplicate-Forward"
      },
	  {
        "source": "Problem/Incident",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "relates to",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "Gantt End to End",
        "target": "System.LinkTypes.Dependency-Forward"
      },
      {
        "source": "Gantt End to Start",
        "target": "System.LinkTypes.Dependency-Forward"
      },
	  {
        "source": "Gantt Start to Start",
        "target": "System.LinkTypes.Related"
      },
      {
        "source": "Gantt Start to End",
        "target": "System.LinkTypes.Dependency-Reverse"
      }
    ]
  },

**Comment by fiolevasy** [comment]
It seems that export tools doesn't support "Blocked by" link.
Jira API returns two type of links: outwardIssue and inwardIssue. "Blocks" links is returned as outwardIssue:
  "id": "30423",
```
                "self": "https://jira_server/jira/rest/api/2/issueLink/30423",
                "type": {
                    "id": "10000",
                    "name": "Blocks",
                    "inward": "is blocked by",
                    "outward": "blocks",
                    "self": "https://jira_server/jira/rest/api/2/issueLinkType/10000"
                },
                "outwardIssue": {
                    "id": "21869",
                    "key": "LAUNC-2",
                    "self": "https://jira_server/jira/rest/api/2/issue/21869",
                    "fields": {


```
And "Blocked by" link is returned as inwardIssue.
```
                "id": "30421",
                "self": "https://jira_server/jira/rest/api/2/issueLink/30421",
                "type": {
                    "id": "10000",
                    "name": "Blocks",
                    "inward": "is blocked by",
                    "outward": "blocks",
                    "self": "https://jira_server/jira/rest/api/2/issueLinkType/10000"
                },
                "inwardIssue": {
                    "id": "13409",
                    "key": "PROJECT-1",
                    "self": "https://jira_server/jira/rest/api/2/issue/13409",
                    "fields": {
                "id": "30421",
                "self": "https://jira_server/jira/rest/api/2/issueLink/30421",
                "type": {
                    "id": "10000",
                    "name": "Blocks",
                    "inward": "is blocked by",
                    "outward": "blocks",
                    "self": "link"
                },
                "inwardIssue": {
```
As I can see from the source 
```
namespace JiraExport
{
    public class JiraItem

        private static List<JiraLink> ExtractLinks(string sourceKey, IEnumerable<JObject> issueLinks)
        {
            var links = new List<JiraLink>();

            foreach (var issueLink in issueLinks)
            {
                var targetIssueKey = issueLink.ExValue<string>("$.outwardIssue.key");
                if (string.IsNullOrWhiteSpace(targetIssueKey))
                    continue;

                var type = issueLink.ExValue<string>("$.type.name");

                var link = new JiraLink() { SourceItem = sourceKey, TargetItem = targetIssueKey, LinkType = type };
                links.Add(link);
            }

            return links;
        }
}
```
only outwardIssue links are parsed.
